http/proxy_server.go

```diff
@@ -0,0 +1,266 @@
+	default:
+		s.serveNonGet(w, r)
+	}
+}
+
+func (s *ProxyServer) serveGet(w http.ResponseWriter, r *http.Request) {
+	// Determine the last write TXID seen by
+	var txid uint64
+	if cookie, _ := r.Cookie(TXIDCookieName); cookie != nil {
+		txid, _ = ltx.ParseTXID(cookie.Value)
+	}
+
+	// No TXID or we couldn't parse it. Just send to the target.
+	if txid == 0 {
+		s.logf("proxy: %s %s: no client txid, proxying to target", r.Method, r.URL.Path)
+		s.proxyToTarget(w, r)
+		return
+	}
+
+	// Lookup our database that we use for TXID tracking.
+	// If the database hasn't been created yet, just send to target.
+	db := s.store.DB(s.DBName)
+	if db == nil {
+		s.logf("proxy: %s %s: no database %q, proxying to target", r.Method, r.URL.Path, s.DBName)
+		s.proxyToTarget(w, r)
+		return
+	}
+
+	// Wait for database to catch up to TXID.
+	ticker := time.NewTicker(s.PollTXIDInterval)
+	defer ticker.Stop()
+
+	ctx, cancel := context.WithTimeout(r.Context(), s.PollTXIDTimeout)
+	defer cancel()
+
+	var pos litefs.Pos
+LOOP:
+	for {
+		if pos = db.Pos(); pos.TXID >= txid {
+			s.logf("proxy: %s %s: database %q at txid %s, proxying to target", r.Method, r.URL.Path, s.DBName, ltx.FormatTXID(pos.TXID))
+			break LOOP
+		}
+
+		select {
+		case <-ctx.Done():
+			s.logf("proxy: %s %s: database %q at txid %s, requires txid %s, proxy timeout", r.Method, r.URL.Path, s.DBName, ltx.FormatTXID(pos.TXID), ltx.FormatTXID(txid))
+			http.Error(w, "Proxy timeout", http.StatusGatewayTimeout)
+			return
+		case <-ticker.C:
+		}
+	}
+
+	// Send request to the target once we've caught up to the last write seen.
+	s.proxyToTarget(w, r)
+}
+
+func (s *ProxyServer) serveNonGet(w http.ResponseWriter, r *http.Request) {
+	// If this is the primary, send the request to the target.
+	if s.store.IsPrimary() {
+		s.logf("proxy: %s %s: node is primary, proxying to target", r.Method, r.URL.Path)
+		s.proxyToTarget(w, r)
+		return
+	}
+
+	// Look up the hostname of the primary. If there's no primary info then
+	// go ahead and send the request
+	info := s.store.PrimaryInfo()
+	if info == nil {
+		s.logf("proxy: %s %s: no primary available, proxying to target", r.Method, r.URL.Path)
+		s.proxyToTarget(w, r)
+		return
+	}
+
+	// If this is a replica, then we'll redirect the request to the primary.
+	w.Header().Set("fly-replay", "instance="+info.Hostname)
+}
+
+func (s *ProxyServer) proxyToTarget(w http.ResponseWriter, r *http.Request) {
+	// Update request URL to target server.
+	r.URL.Scheme = "http"
+	r.URL.Host = s.Target
+
+	resp, err := http.DefaultTransport.RoundTrip(r)
+	if err != nil {
+		http.Error(w, "Proxy error: "+err.Error(), http.StatusBadGateway)
+		return
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// Inject cookie if this is a write.
+	if r.Method != http.MethodGet {
+		if db := s.store.DB(s.DBName); db != nil {
+			pos := db.Pos()
+			s.logf("proxy: %s %s: setting txid cookie to %s", r.Method, r.URL.Path, ltx.FormatTXID(pos.TXID))
+			http.SetCookie(w, &http.Cookie{
```

Comment:
Does this override an existing header on the response or does it append? We'll need it to append.

Reply:
It appends the cookie so it shouldn't affect application cookies (unless they're also named `__txid`): https://cs.opensource.google/go/go/+/refs/tags/go1.20:src/net/http/cookie.go;l=171

Reply:
Just to clarify, `Header().Add()` appends. If it overrode it then it would be `Header().Set()`.

Reply:
I don't see a `Header().Add()` or a `Header().Set()` with regard to this cookie 🤔 But if you say we're good then I'll believe you 😅

Reply:
The `http.SetCookie()` calls `w.Header().Add()`. Sorry, I should have been more clear.

```go
func SetCookie(w ResponseWriter, cookie *Cookie) {
	if v := cookie.String(); v != "" {
		w.Header().Add("Set-Cookie", v)
	}
}
```